src/GraphQL/Extensions/GraphQLBuilderExtensions.cs

```diff
@@ -1168,5 +1168,28 @@ public static IGraphQLBuilder AddExecutionStrategy<TExecutionStrategy>(this IGra
             return builder;
         }
         #endregion
+
+        #region - UseTelemetry -
+#if NET5_0_OR_GREATER
+        /// <summary>
+        /// Configures the GraphQL server to collect traces via the <see cref="System.Diagnostics.Activity">System.Diagnostics.Activity API</see> and records events that match the
+        /// <see href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/graphql.md">OpenTelemetry recommendations</see>.
+        /// Trace data contains the GraphQL operation name, the operation type, and the optionally the document.
+        /// </summary>
+        /// <remarks>
+        /// When applicable, place after calls to UseAutomaticPersistedQueries to ensure that the query document is recorded properly.
+        /// </remarks>
+        public static IGraphQLBuilder UseTelemetry(this IGraphQLBuilder builder, Action<TelemetryOptions>? configure = null)
+            => UseTelemetry(builder, configure != null ? (opts, _) => configure(opts) : null);
+
+        /// <inheritdoc cref="UseTelemetry(IGraphQLBuilder, Action{TelemetryOptions}?)"/>
+        public static IGraphQLBuilder UseTelemetry(this IGraphQLBuilder builder, Action<TelemetryOptions, IServiceProvider>? configure)
+        {
+            builder.Services.Configure(configure);
+            builder.ConfigureExecution<TelemetryProvider>();
```

Comment:
I wanted to add an equivalent comment.

See https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md#writing-a-custom-instrumentation-library:

> The instrumentation libraries requiring state management are usually hard to auto-instrument. Therefore, they take the risk of not being supported by [OpenTelemetry .NET Automatic Instrumentation](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation).

Take notice that the BCL libraries like SqlClient, HttpClient, ASP.NET Core always emit activities.

Reply:
I would like @sungam3r 's opinion on this.  At the moment, adding a single line of code doesn't seem too much to ask for.  However, I'm not familiar with OpenTelemetry Automatic Instrumentation.  And it is true that the effect would be minimal.

The thing is, to make this change would require a different implementation approach entirely: it could not be implemented as an `IConfigureExecution` because then there would be no way to override the implementation with a custom implementation.  It would probably have to integrate more tightly to the `IDocumentExecuter`.  And it might then push these changes back to GraphQL.NET 8.

Reply:
@Shane32, maybe better postponing it to v8 and do it "the right way"? And send people who need the OTel here and now (like me) to this PR as a reference.

Reply:
Hey @Shane32. I just wanted to remind about this @Kielek's comment. The automatic instrumentation was the original reason for OTel support request.

Reply:
Still waiting for @sungam3r 's input here.  My opinion has not changed and is explained above.  If there are any particular reasons why it should be built-in (versus a single line of code to enable), I would very much like to hear it.

Reply:
The automatic instrumentation allows enabling/disabling different instrumentations by configuration and without any code change. Unfortunately, it's in the early development version and I can't use it in production. But this would allow me to drop all the switches I introduced in my code to be able to disable selected instrumentations without code changes. For example, I could just append `GraphQL` to the [environment variable](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/config.md#instrumented-libraries-and-frameworks) to enable it and remove it when I think it's not necessary, too noisy, or buggy
```
OTEL_DOTNET_AUTO_TRACES_ENABLED_INSTRUMENTATIONS=AspNet,HttpClient,Elasticsearch,GraphQL
```
In my opinion, this is a very powerful feature, and I would prefer using it over manual code changes. Just need more libraries to support it out of the box.

Reply:
Makes sense. But then, you can also add the call to AddTelemetry now, and do all the same as you previously described, adding or removing from the environment variable to enable/disable.

Reply:
But then how do you disable it without removing `AddTelemetry`?

Reply:
I think I understand what you mean. `AddTelemetry` will add the `Activity` but won't enable it because I need to call `AddOpenTelemetry()` to make it work.

Reply:
> Still waiting for @sungam3r 's input here. My opinion has not changed and is explained above. If there are any particular reasons why it should be built-in (versus a single line of code to enable), I would very much like to hear it.

The "single line of code to enable" is problematic for OTel .NET auto-instrumentation. We would have to use "bytecode instrumentation". This would force user to use OTel .NET auto-instrumentation as a .NET Profiler (which is not always possible technically, or users may not want to for example because using a .NET profiler adds an performance overhead). Basically the  "bytecode instrumentation" is reJITing the code - so when the .NET runtime is JITing the code we are changing it in fly to add "this single line". This is also quite complex in implementation and maintenance. Eeach time you would change/add/delete possible the "entry points" of GraphQL instrumentation we would need to update our bytecode instrumentation code. The maintenance effort is like 1 day vs many weeks. Not mentioning potential bugs as manipulating user code is always tricky 😉

Reply:
From the GraphQL.NET library’s point of view, including instrumentation by default is equivalent to calling AddTelemetry automatically. If the source is not registered then nothing happens. It seems as though adding or removing the source can be done via environment variables but I am not familiar with this; this would be handled by OpenTelemetry. 

> The "single line of code to enable" is problematic for OTel .NET auto-instrumentation. 

If the user did not add the one line of code. If they did then this is not a problem. 

> Eeach time you would change/add/delete possible the "entry points" of GraphQL instrumentation we would need to update our bytecode instrumentation code.

It would be necessary to add the required service once. It would never be necessary to use byte code instrumentation or modify any other code, and after a single code change, users could enable or disable instrumentation for GraphQL as they please, via any available means within OpenTelemetry - environment variables, app settings, code, I don’t know. 

If this isn’t correct then I still must not understand.

Reply:
> I think I understand what you mean. `AddTelemetry` will add the `Activity` but won't enable it because I need to call `AddOpenTelemetry()` to make it work.

And enable the GraphQL source

Reply:
Ok got it!

Waiting on @sungam3r ’s input.

Reply:
Long read... I'm 70/30 to keep it opt-in for now. I understand that your interest is to seamlessly integrate as many libraries as possible into OTel ecosystem, in particular auto-instrumentation area of it. This is a good goal. Howewer, integrating tracing APIs into codebase is not always as simple as writing one line of code here and there across publishing/consuming classes. This is true for almost any new API, not only tracing. Some settings/options classes, helpers inevitably appear that integrates tracing into the main code base.

Let's look at the question from the other side. The added trace in itself now carries little benefit without adding a resolver-level tracing. I have an example of implementation, we use a resolver-level tracing in some services at work. So...OK. How to add tracing into resolvers? GraphQL.NET provides a concept of `FieldMiddleware` for it. Just write this middleware and incapsulate tracing inside. So what should we do now in case of suggestion to always include support for tracing? Always inject `TelemetryFieldMiddleware` I suppose 😕 .  Do you feel the idea? The instrumented library (GraphQL.NET in our case) already has some design and the authors provided the necessary extension points, which you can integrate to add the necessary functions, including tracing or whatever. These extension points can be 1, 2 or 10. We do not smear the code of generation of activities throughout the code base. Adding new featuress in such a design is always opt-in, at least we follow exactly this approach, guided by the principle of the least surprise. Making integration of tracing opt-out may break expectations.

Let's see how we can provide resolver-level tracing and then decide.

Reply:
> Take notice that the BCL libraries like SqlClient, HttpClient, ASP.NET Core always emit activities.

This is a good reason to do the same, but I still treat this behavior with caution.

Reply:
> The added trace in itself now carries little benefit without adding a resolver-level tracing. I have an example of implementation, we use a resolver-level tracing in some services at work.

I would say the opposite at my work: resolver-level tracing carries little benefit, but we have an extensive system for query-level tracing, such that we can review which queries are being executed which consume the most execution time.  Once the query is identified, it is simple to understand why the query performed poorly without needing the extensive tracing from each resolver -- especially if we are temporarily logging the user id and/or variables provided with the query.

In any case, we can see that different people may have different points of view on this.

Reply:
> Always inject TelemetryFieldMiddleware I suppose

I don't think this assumption is correct. Both of you agreed that resolver telemetry shouldn't be enabled by default, it means the user can't assume it will work by just enabling telemetry manually or automatically. The user will need to read the documentation and perform additional configuration. The same way as sanitizing the document and adding any other enrichment will reacquire delegates configuration.

> The added trace in itself now carries little benefit without adding a resolver-level tracing

The query tracing alone is pretty helpful. It allows following the request flow within the system without zooming in too much. At the end of the day, telemetry is not a profiling tool but observability. If one wants to get more insights like resolver telemetry, he is supposed to perform additional configuration.

Reply:
> Once the query is identified, **it is simple to understand why the query performed poorly**

It depends and experience may vary.

> In any case, we can see that different people may have different points of view on this.

Yep. And because for some of them (like me) resolver tracing is vital I suggest to implement it as opt-in feature - `GraphQLTelemetryMiddleware`. I can post PR targeting this one.

Reply:
> The user will need to read the documentation and perform additional configuration.

Exactly. This is by design I suppose. Telemetry in this regard do not differ from most other features that we added in recent years. See all methods on `IGraphQLBuilder`: serializers, validation rules, complexity, auto schema, APQ, caching and so on. Practically each of these components **may** be plugged in by default when calling `AddGraphQL` in some _disabled_ state controlled by app switches/env vars. Then user may just set variable/config value and viola! But we do not use such a design and is unlikely to start using (including due to problems with performance hit). Waiting on @Shane32 ’s input.

Reply:
A few comments:

1. Adding middleware always, while CERTAINLY not preferable, can be designed to as to not introduce any new allocations (when no listeners are attached) because of the ValueTask pipeline.  JIT would likely optimize the code in such a manner that would make it perform almost as well as if the middleware were not installed.

2. Microsoft's best practices discourage capturing a 'large volume of events' with `Activity`, and suggest using `ILogger` instead for this purpose.  Personally, I would take this to include resolver-level tracing (but okay with opt-in if users request it).

https://learn.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs#best-practices-3

> Events are stored in an in-memory list until they can be transmitted which makes this mechanism only suitable for recording a modest number of events. For a large or unbounded volume of events, it's better to use a logging API focused on this task, such as [ILogger](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/). ILogger also ensures that the logging information will be available regardless whether the app developer opts to use distributed tracing. ILogger supports automatically capturing the active Activity IDs so messages logged via that API can still be correlated with the distributed trace.

Reply:
@sungam3r I'm not exactly sure what input you were waiting on here on this long thread.... at 4am I don't feel like reading the whole thing just now!  If you're wondering if I suggest opt-in to telemetry or on-by-default, I would normally suggest on-by-default to match other of MS' code......... however, to be able to customize it would require extra interfaces and DI configuration.  So, I say to make it opt-in for now.  Perhaps in a later version of GraphQL.NET we make it on-by-default.

I also think we should consider adding logging via ILogger, just as other of MS' libraries do.  We would need a dependency, so maybe this should go in another nuget package like GraphQL.Logging.

Reply:
> So, I say to make it opt-in for now. Perhaps in a later version of GraphQL.NET we make it on-by-default.

So do I. Let's consider this a consensus so far.

> Microsoft's best practices discourage capturing a 'large volume of events' with Activity, and suggest using ILogger instead for this purpose. Personally, I would take this to include resolver-level tracing (but okay with opt-in if users request it).

Either traces or logs (or both) may be used for resolver level. The advantage of tracing is that there are ready-made visualization tools that reduce time to understand the general picture of delays handling typical http/graphql request. Logs in this regard is a plain stream of data. I understand the note about 'large volume of events'. This is not a problem when properly configured, for example, only top-level fields may be traced or all fields with depth < 3, or explicitly specified set of fields. This is a matter of available configuration options and sane default values.

> I also think we should consider adding logging via ILogger, just as other of MS' libraries do. We would need a dependency, so maybe this should go in another nuget package like GraphQL.Logging.

I would move this one out of scope for now.